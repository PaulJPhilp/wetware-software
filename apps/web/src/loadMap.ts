import { MapDataSchema, type MapData } from "./schema";

export function loadMap(input: string | unknown): MapData {
  let data: unknown;

  if (typeof input === "string") {
    try {
      data = JSON.parse(input);
    } catch (error) {
      throw new Error(`Invalid JSON: ${error instanceof Error ? error.message : "Unknown error"}`);
    }
  } else {
    data = input;
  }

  const result = MapDataSchema.safeParse(data);
  if (!result.success) {
    const issues = result.error.issues
      .map((issue) => `${issue.path.join(".")}: ${issue.message}`)
      .join("; ");
    throw new Error(`Validation failed: ${issues}`);
  }

  return result.data;
}
