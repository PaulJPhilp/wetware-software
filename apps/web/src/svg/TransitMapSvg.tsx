import { useEffect, useRef } from "react";
import type { MapData } from "../schema";
import type { Theme } from "../theme";
import { defaultTheme } from "../theme";
import { buildPolylinePoints } from "./helpers/geometry";
import { calculateLabelPosition } from "./helpers/label";

interface TransitMapSvgProps {
  data: MapData;
  theme?: Partial<Theme>;
  onMount?: (el: SVGSVGElement) => void;
}

export function TransitMapSvg({ data, theme = {}, onMount }: TransitMapSvgProps) {
  const mergedTheme = { ...defaultTheme, ...theme };
  const svgRef = useRef<SVGSVGElement>(null);

  useEffect(() => {
    if (svgRef.current && onMount) {
      onMount(svgRef.current);
    }
  }, [onMount]);

  // Calculate viewBox
  const allX = data.stops.map((s) => s.x);
  const allY = data.stops.map((s) => s.y);
  const minX = Math.min(...allX) - 20;
  const maxX = Math.max(...allX) + 120; // extra for labels
  const minY = Math.min(...allY) - 20;
  const maxY = Math.max(...allY) + 20;
  const width = maxX - minX;
  const height = maxY - minY;

  // Helper to check if stop is interchange
  const isInterchange = (stopId: string) => {
    return data.interchanges?.some((interchange) => interchange.stopIds.includes(stopId)) ?? false;
  };

  // Create stop map for quick lookup
  const stopMap = new Map(data.stops.map((stop) => [stop.id, stop]));

  return (
    <svg
      ref={svgRef}
      viewBox={`${minX} ${minY} ${width} ${height}`}
      className="transit-map-svg"
      aria-label="Transit map"
      style={{ backgroundColor: mergedTheme.colors.background }}
    >
      <title>Transit map</title>
      {/* Render lines */}
      {data.lines.map((line) => {
        const stops = line.stops
          .map((id) => stopMap.get(id))
          .filter((s): s is (typeof data.stops)[number] => Boolean(s));
        if (stops.length < 2) return null;
        const points = buildPolylinePoints(stops);
        return (
          <polyline
            key={line.id}
            points={points}
            stroke={line.color}
            strokeWidth={mergedTheme.lineWidth}
            fill="none"
          />
        );
      })}

      {/* Render stations */}
      {data.stops.map((stop) => {
        const isInterchangeStop = isInterchange(stop.id);
        const radius = isInterchangeStop
          ? mergedTheme.interchangeRadius
          : mergedTheme.stationRadius;
        const labelPosition = calculateLabelPosition(
          stop,
          data.stops,
          data.interchanges || [],
          mergedTheme,
        );
        return (
          <g
            key={stop.id}
            className="station"
            role="button"
            tabIndex={0}
            aria-label={`${stop.title} (${stop.years})`}
          >
            <circle
              cx={stop.x}
              cy={stop.y}
              r={radius}
              fill={mergedTheme.colors.text}
              stroke={isInterchangeStop ? mergedTheme.colors.background : undefined}
              strokeWidth={isInterchangeStop ? 2 : 0}
            />
            <text
              x={labelPosition.x}
              y={labelPosition.y}
              fontSize={mergedTheme.fontSize}
              fill={mergedTheme.colors.text}
              textAnchor={labelPosition.textAnchor as import("./helpers/label").TextAnchor}
              dominantBaseline={
                labelPosition.dominantBaseline as import("./helpers/label").DominantBaseline
              }
            >
              {stop.title}
            </text>
            <title>{`${stop.title} (${stop.years})`}</title>
          </g>
        );
      })}
    </svg>
  );
}
