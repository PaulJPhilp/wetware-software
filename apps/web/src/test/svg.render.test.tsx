import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { loadMap } from "../loadMap";
import type { MapData } from "../schema";
import { TransitMapSvg } from "../svg/TransitMapSvg";

describe("TransitMapSvg", () => {
  const testData: MapData = loadMap({
    stops: [
      {
        id: "stop1",
        title: "Station One",
        years: "2020-2025",
        x: 100,
        y: 200,
        bullets: ["Bullet 1"],
        carryForward: undefined,
      },
      {
        id: "stop2",
        title: "Station Two",
        years: "2021-2025",
        x: 150,
        y: 250,
        bullets: ["Bullet A"],
        carryForward: undefined,
      },
      {
        id: "stop3",
        title: "Station Three",
        years: "2022-2025",
        x: 200,
        y: 300,
        bullets: ["Bullet X"],
        carryForward: undefined,
      },
    ],
    lines: [
      {
        id: "line1",
        name: "Red Line",
        color: "#ff0000",
        stops: ["stop1", "stop2"],
      },
      {
        id: "line2",
        name: "Blue Line",
        color: "#0000ff",
        stops: ["stop2", "stop3"],
      },
    ],
    interchanges: [
      {
        stopIds: ["stop2"],
      },
    ],
  } as unknown as ReturnType<typeof loadMap>);
  it("renders the correct number of polylines", () => {
    const { container } = render(<TransitMapSvg data={testData} />);
    const polylines = container.querySelectorAll("polyline");
    expect(polylines).toHaveLength(2);
  });

  it("renders the correct number of stations", () => {
    const { container } = render(<TransitMapSvg data={testData} />);
    const circles = container.querySelectorAll("circle");
    expect(circles).toHaveLength(3);
  });

  it("renders interchange station with larger radius", () => {
    const { container } = render(<TransitMapSvg data={testData} />);
    const circles = container.querySelectorAll("circle");
    // stop2 is interchange
    const interchangeCircle = Array.from(circles).find(
      (circle) => circle.getAttribute("cx") === "150"
    );
    expect(interchangeCircle?.getAttribute("r")).toBe("12"); // interchangeRadius
    expect(interchangeCircle?.getAttribute("stroke")).toBe("white"); // background color
  });

  it("renders regular stations with normal radius", () => {
    const { container } = render(<TransitMapSvg data={testData} />);
    const circles = container.querySelectorAll("circle");
    const regularCircles = Array.from(circles).filter(
      (circle) => circle.getAttribute("cx") !== "150"
    );
    regularCircles.forEach((circle) => {
      expect(circle.getAttribute("r")).toBe("8"); // stationRadius
      expect(circle.getAttribute("stroke")).toBeNull();
    });
  });

  it("includes title elements for accessibility", () => {
    const { container } = render(<TransitMapSvg data={testData} />);
    const titles = container.querySelectorAll("title");
    expect(titles).toHaveLength(3);
    expect(titles[0]?.textContent).toBe("Station One (2020-2025)");
    expect(titles[1]?.textContent).toBe("Station Two (2021-2025)");
    expect(titles[2]?.textContent).toBe("Station Three (2022-2025)");
  });

  it("renders labels with collision avoidance", () => {
    const { container } = render(<TransitMapSvg data={testData} />);
    const texts = container.querySelectorAll("text");
    expect(texts).toHaveLength(3);
    // Check that labels are positioned (basic check)
    texts.forEach((text) => {
      expect(text.getAttribute("x")).toBeDefined();
      expect(text.getAttribute("y")).toBeDefined();
    });
  });

  it("respects explicit labelPos", () => {
    const dataWithLabelPos = {
      ...testData,
      stops: testData.stops.map((stop) =>
        stop.id === "stop1" ? { ...stop, labelPos: "top" as const } : stop
      ),
    };
    const { container } = render(<TransitMapSvg data={dataWithLabelPos} />);
    const text = container.querySelector("text");
    expect(text?.getAttribute("text-anchor")).toBe("middle");
    expect(text?.getAttribute("dominant-baseline")).toBe("alphabetic");
  });

  it("matches snapshot", () => {
    const { container } = render(<TransitMapSvg data={testData} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
