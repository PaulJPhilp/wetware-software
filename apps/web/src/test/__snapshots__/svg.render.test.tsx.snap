// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransitMapSvg > matches snapshot 1`] = `
<svg
  class="transit-map-svg"
  style="background-color: white;"
  viewBox="80 180 240 140"
>
  <polyline
    fill="none"
    points="100,200 150,250"
    stroke="#ff0000"
    stroke-width="4"
  />
  <polyline
    fill="none"
    points="150,250 200,300"
    stroke="#0000ff"
    stroke-width="4"
  />
  <g
    class="station"
  >
    <circle
      cx="100"
      cy="200"
      fill="black"
      r="8"
      stroke-width="0"
    />
    <text
      dominant-baseline="middle"
      fill="black"
      font-size="12"
      text-anchor="start"
      x="112"
      y="204"
    >
      Station One
    </text>
    <title>
      Station One (2020-2025)
    </title>
  </g>
  <g
    class="station"
  >
    <circle
      cx="150"
      cy="250"
      fill="black"
      r="12"
      stroke="white"
      stroke-width="2"
    />
    <text
      dominant-baseline="middle"
      fill="black"
      font-size="12"
      text-anchor="start"
      x="166"
      y="254"
    >
      Station Two
    </text>
    <title>
      Station Two (2021-2025)
    </title>
  </g>
  <g
    class="station"
  >
    <circle
      cx="200"
      cy="300"
      fill="black"
      r="8"
      stroke-width="0"
    />
    <text
      dominant-baseline="middle"
      fill="black"
      font-size="12"
      text-anchor="start"
      x="212"
      y="304"
    >
      Station Three
    </text>
    <title>
      Station Three (2022-2025)
    </title>
  </g>
</svg>
`;
